[HUDI-7652] Add new `HoodieMergeKey` API to support simple and composite keys

codope opened the pull request:
### Change Logs

This PR introduces a new class hierarchy for handling merge keys in a more flexible and decoupled manner. It adds the `HoodieMergeKey` interface, along with two implementations: `HoodieSimpleMergeKey` and `HoodieCompositeMergeKey`. This design allows us to extend key-based merge strategies easily.

**Motivation**

The need for introducing a new merge key handling mechanism was driven by the requirement to support different types of keys (simple and complex) without overloading the existing HoodieKey class, which is central to the write path. By segregating merge key handling into its own hierarchy, we avoid potential conflicts and keep modifications localised, improving the maintainability of the code.

**Changes**

1. `HoodieMergeKey`: New API to ensure consistent handling including simple keys and composite keys. It includes methods for retrieving the key and partition path.
2. `HoodieSimpleMergeKey`: Wraps `HoodieKey` and implements the `HoodieMergeKey` interface for simple scenarios where the key is a string.
3. `HoodieCompositeMergeKey`: Implements the  `HoodieMergeKey` interface but allows for complex types as keys, enhancing flexibility for scenarios where a simple string key is not sufficient.
4. `HoodieMergeKeyBasedRecordMerger`: A new implementation of `HoodieRecordMerger` based on `HoodieMergeKey`. If the merge keys are of type `HoodieCompositeMergeKey`, then it returns the older and newer records. Otherwise, it calls the merge method from the parent class.
5. `HoodieMergedLogRecordScanner`: Changes to merge based on `HoodieMergeKey`.
6. Unit tests for the new merger.

These changes do not affect existing functionalities that do not rely on merge keys. It introduces additional classes that are used explicitly for new functionalities involving various key types in merging operations. This ensures minimal to no risk for existing processes.

### Impact

Enhancing the flexibility and robustness of our key-based merge strategies. It helps in keeping our codebase scalable and maintainable, allowing easy extensions and modifications in the future.

### Risk level (write none, low medium or high below)

low

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change. If not, put "none"._

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -243,34 +250,57 @@ public static HoodieMergedLogRecordScanner.Builder newBuilder() {
+        HoodieRecord<T> combinedRecord = (HoodieRecord<T>) recordMerger.merge(prevRecord, readerSchema,
+            newRecord, readerSchema, this.getPayloadProps()).get().getLeft();
+        // If pre-combine returns existing record, no need to update it
+        if (combinedRecord.getData() != prevRecord.getData()) {
```
I don't think we should put any regular code path into a catch block.

codope replied:
Basically, want to fallback to original method if a merger is used that does not support MergeKey based merging. Wanted to avoid another boolean flag in the scanner so as not to expose the implementation detail here. Let me think if there can be a better way to do this.

codope replied:
This code is not obsolete. Have changed as per our offline dicussion.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -81,7 +83,7 @@ public class HoodieMergedLogRecordScanner extends AbstractHoodieLogRecordReader
   public final HoodieTimer timer = HoodieTimer.create();
   // Map of compacted/merged records
-  private final ExternalSpillableMap<String, HoodieRecord> records;
+  private final ExternalSpillableMap<HoodieMergeKey, HoodieRecord> records;
```
If we make the key type as a `Serializable`, is it possible to eliminate the `HoodieMergeKey` abstraction? Because the new change introduces the performace regression on non-SI scenarios.

codope replied:
We did nto want to overload the `HoodieKey`. This was discussed in https://github.com/apache/hudi/pull/10814#pullrequestreview-2003629004

codope replied:
I don't think there is a major perf regression on non-SI path. We are creating only new objects, not doing any further compute.

danny0405 replied:
> we do a separate class hierarchy and not overload HoodieKey

Not sure about the specific background here, but make the `ExternalSpillableMap` key as `Serializable` does not overload anything?

codope replied:
You're right, it does not overload anything. But, if we make it `Serializable`, then we need to make changes at the callers of `HoodieMergedLogRecordScanner.getRecords()` to decide how to change the `Serializable` to String key. A separate abstraction is cleaner as we don't leak any information at layers above.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieSimpleMergeKey.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Wraps {@link HoodieKey} and implements the {@link HoodieMergeKey} interface for simple scenarios where the key is a string.
+ */
+public class HoodieSimpleMergeKey implements HoodieMergeKey {
```
Actually we might could not call this a "simple" merge key, becase the it looks like the recordk key here can take multiple fields in table schema.

codope replied:
record key fields can be multiple but it's still a string combined.

danny0405 replied:
I'm talking about the notion instead of the physical impl.

codope replied:
Ok, would renaming to `HoodiePrimaryMergeKey` be more suitable, more like how database have primary and composite key? The basic idea is that this merge key is based on record key field(s) which will be unique.

codope replied:
This new api is no longer needed.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieMergeKey.java:
```diff
@@ -0,0 +1,48 @@
+ * Defines a standard for all merge keys to ensure consistent handling including simple keys and composite keys.
+ * It includes methods for retrieving the key and partition path.
+ */
+public interface HoodieMergeKey extends Serializable {
```
So this API is mainy introduced to let the key type of log reader mapping cache be generic?

codope replied:
Yes, that's right.

codope replied:
> For SI merging, is the partition path a required param?

This API will not be limited to SI. It is possible that we get same key under two different partition path.

codope commented:
@danny0405 Based on our discussion, I have removed the `HoodieMergeKey` API and created a subclass `HoodieMetadataMergedLogRecordScanner` which works with `ExternalSpillableMap<Serializable, HoodieRecord>` (instead of string keys). It is used only for `HoodieMetadataLogRecordReader` and introduced a `HoodieMetadataRecordMerger`. The merger currently mimics the super class, but it will change for secondary index in subsequent PR. Please review this PR again.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -91,7 +91,7 @@ public class HoodieMergedLogRecordScanner extends AbstractHoodieLogRecordReader
   @SuppressWarnings("unchecked")
-  private HoodieMergedLogRecordScanner(HoodieStorage storage, String basePath, List<String> logFilePaths, Schema readerSchema,
+  HoodieMergedLogRecordScanner(HoodieStorage storage, String basePath, List<String> logFilePaths, Schema readerSchema,
                                        String latestInstantTime, Long maxMemorySizeInBytes,
```
Fix the indentation and make the constructor scope as protected.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -271,6 +266,16 @@ public <T> void processNextRecord(HoodieRecord<T> newRecord) throws IOException
     }
   }
 
+  static <T> HoodieRecord getLatestHoodieRecord(HoodieRecord<T> newRecord, HoodieRecord<T> combinedRecord, String key) {
```
make the scope as protected.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMetadataMergedLogRecordScanner.java:
```diff
@@ -0,0 +1,308 @@
+        isBitCaskDiskMapCompressionEnabled, withOperationField, false, partitionName, internalSchema, keyFieldOverride, enableOptimizedLogBlocksScan, recordMerger,
+        hoodieTableMetaClientOption);
+    try {
+      // Store merged records for all versions for this log file, set the in-memory footprint to maxInMemoryMapSize
```
If we make the super class `HoodieMergedLogRecordScanner` records cache key as generic type, we can actually eliminate the redundant initialization of the map.

codope replied:
Done.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMetadataMergedLogRecordScanner.java:
```diff
@@ -0,0 +1,308 @@
+
+  @Override
+  public <T> void processNextRecord(HoodieRecord<T> newRecord) throws IOException {
+    // Merge the new record with the existing record in the map
```
Can overcome the duplicate code if we make the cache key as generic type.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMetadataMergedLogRecordScanner.java:
```diff
@@ -0,0 +1,308 @@
+  }
+
+  @Override
+  public Map<String, HoodieRecord> getRecords() {
```
Can overcome the duplicate code if we make the cache key as generic type.

danny0405 reviewed (commented):
Thanks for the contribution, I have left some comments.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -222,7 +223,8 @@ public Iterator<HoodieRecord> iterator() {
 
   public Map<String, HoodieRecord> getRecords() {
-    return records;
+    return records.entrySet().stream().collect(
```
Can we override this method instead?

codope replied:
Since we have made the super class `records` cache generic, so we have to make changes here to return String.

danny0405 replied:
Does generic type solves the problem? Or can we do a implicit type cast.

codope replied:
No type cast is required, generic type is sufficient. So, this method should always return string otherwise we have to change all the callers, including merge handle, which we don't want to do.

danny0405 replied:
> generic type is sufficient

For generic type, I mean to make the it as a Java generic type like `K extends Serializable` instead of just Serializable, does that work?

codope replied:
That will also work but let me put down my thoughts why I considered to use `Serializable` as the key type directly instead of a generic type `K extends Serializable`. Firstly, if we use `K extends Serializable`, the creators of `HoodieMergedLogRecordScanner` need to specify the correct type that matches the key. This would increase the surface area of the changes.

Plus, using `Serializable` as the key type directly is more flexible as it allows the `ExternalSpillableMap` to accept any object that implements Serializable. This means you can use a wide variety of types as keys without needing to specify a particular type parameter when instantiating the map.

One drawback of using `Serializable` directly as the key type is the reduced type safety. However, this is only an internal API implementation so it is less prone to being misused.

codope replied:
Discussed offline, and we are aligned on the following approach (which is also done as part of the last commit):
Introduce a new abstract class `BaseHoodieMergedLogRecordScanner` which subclasses from `AbstractHoodieLogRecordReader`. The new abstract class holds the `records` map as `ExternalSpillableMap<K, HoodieRecord>` and exposes `public abstract Map<K, HoodieRecord> getRecords()` API. The existing `HoodieMergedLogRecordScanner` now derives from the new abstract class (instead of `AbstractHoodieLogRecordReader`) and uses String keys.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataLogRecordReader.java:
```diff
@@ -150,9 +151,9 @@ private static Predicate<String> createPrefixMatchingPredicate(List<String> keyP
-        new HoodieMergedLogRecordScanner.Builder()
-            .withKeyFiledOverride(HoodieMetadataPayload.KEY_FIELD_NAME)
+    private final HoodieMetadataMergedLogRecordScanner.Builder scannerBuilder =
+        new HoodieMetadataMergedLogRecordScanner.Builder()
```
Do we need the scanner for all kinds of metadata index types? The full outer join may bring in perforamce regression.

codope replied:
No, we need only for secondary index. I have changes in a subsequent PR. But, i'll do as suggested in this PR itself.

codope commented on hudi-cli/src/main/java/org/apache/hudi/cli/commands/HoodieLogFileCommand.java:
```diff
@@ -225,7 +225,7 @@ storage, new StoragePath(logFilePathPattern)).stream()
       System.out.println("===========================> MERGING RECORDS <===================");
       HoodieMergedLogRecordScanner scanner =
-          HoodieMergedLogRecordScanner.newBuilder()
+          (HoodieMergedLogRecordScanner) HoodieMergedLogRecordScanner.newBuilder()
```
Note for reviewers: Need to cast because I moved the builder to the base class to enahnce reuse.

danny0405 replied:
Let's add the builder impl for each sub-class, just like what we do to original `HoodieUnMergedLogRecordScanner` and `HoodieMergedLogRecordScanner`.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMetadataMergedLogRecordScanner.java:
```diff
@@ -0,0 +1,133 @@
+/**
+ * Merged log record scanner for metadata table using {@link HoodieMetadataRecordMerger}.
+ */
+public class HoodieMetadataMergedLogRecordScanner extends BaseHoodieMergedLogRecordScanner<String> {
```
So in this patch, we use `String` key for functional index? Should we make it `Serializable` instead?

codope replied:
We need Serializable for sec index which I will do in #11146 after landing this patch.
